example-widget-mui/src/UploadImagePage/UploadImagePage.tsx

```diff
@@ -0,0 +1,245 @@
+ * limitations under the License.
+ */
+
+import { MuiCapabilitiesGuard } from '@matrix-widget-toolkit/mui';
+import { useWidgetApi } from '@matrix-widget-toolkit/react';
+import ImageIcon from '@mui/icons-material/Image';
+import UploadIcon from '@mui/icons-material/Upload';
+import {
+  Alert,
+  AlertTitle,
+  Box,
+  Button,
+  Card,
+  CardContent,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogContentText,
+  DialogTitle,
+  Grid,
+  Input,
+  Typography,
+} from '@mui/material';
+import { unstable_useId as useId } from '@mui/utils';
+import {
+  EventDirection,
+  WidgetApiFromWidgetAction,
+  WidgetEventCapability,
+} from 'matrix-widget-api';
+import { ChangeEvent, ReactElement, useCallback, useState } from 'react';
+import { NavigationBar } from '../NavigationPage';
+import { StoreProvider } from '../store';
+import { ImageListView } from './ImageListView';
+
+import { ROOM_EVENT_UPLOADED_IMAGE, UploadedImageEvent } from '../events';
+
+/**
+ * A component that showcases how to upload image files and render them in a widget.
+ */
+export const UploadImagePage = (): ReactElement => {
+  const widgetApi = useWidgetApi();
+  const [errorDialogOpen, setErrorDialogOpen] = useState<boolean>(false);
+  const [selectedFile, setSelectedFile] = useState<File | null>(null);
+  const [loading, setLoading] = useState<boolean>(false);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const titleId = useId();
+  const descriptionId = useId();
+
+  const handleFileChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      const file = event.target.files?.[0] || null;
+      if (file) {
+        setSelectedFile(file);
+      }
+    },
+    [],
+  );
+
+  const handleFileUpload = useCallback(() => {
+    const uploadImage = async () => {
+      if (selectedFile) {
+        if (!selectedFile.type.startsWith('image/')) {
+          setErrorMessage(
+            'Please select a valid image file. You can upload any image format that is supported by the browser.',
+          );
+          setErrorDialogOpen(true);
+          return;
+        }
+
+        try {
+          const getConfigMedia = (await widgetApi.getMediaConfig())[
+            'm.upload.size'
+          ];
+          if (getConfigMedia && getConfigMedia < selectedFile.size) {
+            setErrorMessage(
+              `the max file size ${selectedFile.size} should not be more then ${getConfigMedia}.`,
+            );
+            setErrorDialogOpen(true);
+            setSelectedFile(null);
+            return;
+          } else {
+            setLoading(true);
+            const responseUploadMedia =
+              await widgetApi.uploadFile(selectedFile);
+            const url = getHttpUriForMxc(
+              responseUploadMedia.content_uri,
+              widgetApi.widgetParameters.baseUrl,
+            );
+            if (url) {
+              await widgetApi.sendRoomEvent<UploadedImageEvent>(
```

Comment:
In my opinion we should post the MXC URL with the event here instead.

Reply:
I've updated with this change.